lib/babe/babe.go

```diff
@@ -219,15 +227,26 @@ func (b *Session) setAuthorityIndex() error {
 	return fmt.Errorf("key not in BABE authority data")
 }
 
+func isClosed(ch <-chan struct{}) bool {
+	select {
+	case <-ch:
+		return true
+	default:
+	}
+
+	return false
+}
+
 func (b *Session) checkForKill() {
-	<-b.kill
-	b.stop()
+	if isClosed(b.kill) {
+		b.stop()
+	}
 }
 
 func (b *Session) isClosed() bool {
 	b.lock.Lock()
 	defer b.lock.Unlock()
-	return b.closed
+	return atomic.LoadUint32(&b.closed) == uint32(0)
```

Comment:
same here

Reply:
also, do we need a lock still if this is atomic?

Reply:
I updated this to started and removed the lock, please re-check.